server/src/main/java/com/broadcom/lsp/cobol/service/CopybookServiceImpl.java

```diff
@@ -47,77 +51,95 @@
+      FileSystemService files,
+      @Named("CACHE-MAX-SIZE") int cacheSize,
+      @Named("CACHE-DURATION") int duration,
+      @Named("CACHE-TIME-UNIT") String timeUnitName) {
     this.dataBus = dataBus;
     this.settingsService = settingsService;
     this.files = files;
-
-    dataBus.subscribe(REQUIRED_COPYBOOK_EVENT, this);
+    copybookCache = CacheBuilder.newBuilder()
+        .expireAfterWrite(duration, TimeUnit.valueOf(timeUnitName))
+        .maximumSize(cacheSize)
+        .build();
     dataBus.subscribe(ANALYSIS_FINISHED_EVENT, this);
   }
 
   @Override
-  public void invalidateURICache() {
-    copybookPaths.clear();
+  public void invalidateCache() {
     copybooksForDownloading.clear();
+    copybookCache.invalidateAll();
   }
 
   /**
-   * Depends on DataEvent type it will be handled with appropriate handler: {@link
-   * #handleRequiredCopybookEvent} or {@link #handleAnalysisFinishedEvent}.
+   * Retrieve and return the copybook by its name. Copybook may cached to limit interactions with
+   * the file system.
+   *
+   * Resolving works in synchronous way. Resolutions with different copybook names will not block each other.
    *
-   * @param event the instance of {@link RequiredCopybookEvent} or {@link AnalysisFinishedEvent}
+   * @param copybookName - the name of the copybook to be retrieved
+   * @param documentUri - the currently processing document that contains the copy statement
+   * @param copybookProcessingMode - text document synchronization type
+   * @return a CopybookModel that contains copybook name, its URI and the content
    */
+  public CopybookModel resolve(
+      @Nonnull String copybookName,
+      @Nonnull String documentUri,
+      @Nonnull CopybookProcessingMode copybookProcessingMode) {
+    try {
+      return copybookCache.get(copybookName, () -> resolveSync(copybookName, documentUri, copybookProcessingMode));
+    } catch (ExecutionException e) {
+      LOG.error("Can't resolve copybook '{}'.", copybookName, e);
+      return new CopybookModel(copybookName, null, null);
+    }
+  }
+
   @Override
-  public void observerCallback(DataEvent event) {
-    if (event instanceof RequiredCopybookEvent) {
-      handleRequiredCopybookEvent((RequiredCopybookEvent) event);
-    } else if (event instanceof AnalysisFinishedEvent) {
-      handleAnalysisFinishedEvent((AnalysisFinishedEvent) event);
+  public void store(CopybookModel copybookModel) {
+    copybookCache.put(copybookModel.getName(), copybookModel);
+  }
+
+  private CopybookModel resolveSync(
+      @Nonnull String copybookName,
+      @Nonnull String documentUri,
+      @Nonnull CopybookProcessingMode copybookProcessingMode) throws ExecutionException, InterruptedException {
+    String cobolFileName = files.getNameFromURI(documentUri);
+    String uri = retrieveURI(
+        settingsService.getConfiguration(COPYBOOK_RESOLVE.label, cobolFileName, copybookName).get());
+    if (uri.isEmpty()) {
+      if (copybookProcessingMode.download) {
```

Comment:
Seems like it may be collapsed

Reply:
I can't collapse it. Here I need to manage three cases:
URI is present => read copybook from filesystem
URI is empty and we downloading (DidOpen case) => add copybook for downloading, return stub.
URI is empty and we don't download (DidChange case) => just return stub.

Reply:
Yeah, you are right